plenum/server/consensus/consensus_data_provider.py

```diff
@@ -48,22 +60,85 @@ def is_primary(self) -> bool:
     def preprepared(self) -> List[PrePrepare]:
         """
         List of PrePrepare messages, for which quorum of Prepare messages is not reached yet
-        TODO: Make it simple field?
         """
-        return []
+        return self._preprepared
 
     @property
     def prepared(self) -> List[PrePrepare]:
         """
         List of PrePrepare messages, for which quorum of Prepare messages is reached
-        TODO: Make it simple field?
         """
-        return []
+        return self._prepared
 
     @property
     def stable_checkpoint(self) -> int:
-        return 0
+        return self._stable_checkpoint
 
     @property
     def checkpoints(self) -> List[Checkpoint]:
-        return []
+        """
+        List of Checkpoints, which has been sent, but not stabilized yet
+        """
+        return self._checkpoints.values()
+
+    def preprepare_batch(self, pp: PrePrepare):
+        """
+        After pp had validated, it placed into _preprepared list
+        """
+        self._preprepared.append(pp)
+
+    def prepare_batch(self, pp: PrePrepare):
+        """
+        After prepared certificate for pp had collected,
+        it removed from _preprepared and placed into _prepared list
+        """
+        if pp is None:
+            raise LogicError('pp does not exist')
```

Comment:
Actually passing None here is violation of contract imposed by type annotations, I don't think we need an explicit check here

Reply:
ok